File: lib/fc/allocator.rb
# coding: utf-8

module Fc

  def self.calc_live_range( lmd )
    # ラベルの収集
    labels = Hash.new # labels[ラベル] = op番号
    lmd.ops.each_with_index do |op,i|
      labels[op[1]] = i if op[0] == :label
    end

    # 変数の定義・仕様、制御フローグラフの集計
    use_define = Hash.new {|h,k| h[k] = [[],[]] } # 変数ごとの use/define を記録する
    flow = [] # flow[ジャンプ元op番号] = [ジャンプ先のop番号, ... ]
    lmd.ops.each_with_index do |op,i|
      uses = []
      defines = []
      node = []
      case op[0]
      when :label, :asm, :push_result, :push_fastcall_result
        # DO NOTHING
      when :'if'
        uses << op[1]
        node << labels[op[2]]
      when :jump
        node << labels[op[1]]
      when :return
        uses << op[1]
      when :push_arg, :push_fastcall_arg
        uses << op[2]
      when :load, :uminus, :not, :sign_extension, :ref, :call, :fastcall
        defines << op[1]
        uses << op[2]
      when :add, :sub, :and, :or, :xor, :mul, :div, :mod, :eq, :lt, :shift_left, :shift_right, :index, :pget
        defines << op[1]
        uses << op[2] << op[3]
      when :pset
        uses << op[1] << op[2]
      else
        #:nocov:
        raise "invalid op #{op}"
        #:nocov:
      end

      flow << node

      defines.each do |v|
        use_define[v][0] << i if v and v.kind == :local
      end

      uses.each do |v|
        use_define[v][1] << i if v and v.kind == :local
      end
    end

    # 引数は、最初に定義されているものとする
    use_define.each do |v,info|
      info[0] << 0 if v.opt[:local_type] == :arg
    end

    # live range を求める
    lrc = LiveRangeCalculator.new( flow )
    use_define.each do |v,info|
      v.live_range = lrc.calc_live_range( info )
    end
    
  end

  # 
  # 変数の address, location, unuse が設定される
  def self.allocate_register( lmd )
    calc_live_range( lmd )

    # ref(&演算子)を受けた変数を集める
    refered = Hash.new # 変数がrefを受けたかどうか(refされるならメモリは独立で確保しなくてはいけないため)
    lmd.ops.each do |op|
      refered[op[2]] = true if op[0] == :ref
    end

    # live range を求め、{callをまたぐ|refを受ける|引数}だったら、フレームスタック上に確保する
    frame_size = lmd.type.base.size # 帰り値分を予約しておく
    register_vars = Hash.new # レジスタに割り当てる変数, register_vars[変数:Value] = liverange:Range
    lmd.vars.each do |v|
      beyond_call = false
      if v.live_range
        ((v.live_range.min+1)..(v.live_range.max-1)).each do |i|
          beyond_call = true if lmd.ops[i] and lmd.ops[i][0] == :call and not lmd.ops[i][2].type.fastcall?
        end
      end
      
      if v.opt[:local_type] == :result
        # 返り値
        lmd.result.location = lmd.type.fastcall? ? :fastcall_reg : :frame
        lmd.result.address = 0 
      elsif beyond_call or v.opt[:local_type] == :arg or refered[v]
        # 引数か、関数をまたいでいるなら、フレームに割り当てる
        v.address = frame_size
        v.location = lmd.type.fastcall? ? :fastcall_reg : :frame
        frame_size += v.type.size
      elsif v.live_range
        # それ以外の使われてる変数は、レジスターメモリに割り当てる
        register_vars[v] = { live_range: v.live_range }
      else
        # 未使用フラグをたてる
        v.location = :none
        v.unuse = true
      end
    end

    allocate_cond( lmd, register_vars )
    allocate_a( lmd, register_vars )

    # 各ジスタのアドレスを割り当てる
    reg_size = 0
    allocator = Allocator.new register_vars
    allocator.regs.each do |reg|
      # アドレスを算出する
      raise CompileError.new("frame size over on #{lmd}") if reg_size > 16
      # 割り当てる
      reg[:vars].each do |v|
        if lmd.type.fastcall?
          raise CompileError.new("frame size over on #{lmd}") if frame_size + reg_size > 16
          v.location = :fastcall_reg
          v.address = frame_size + reg_size
        else
          v.location = :reg
          v.address = reg_size
        end
      end
      reg_size += 2
    end

    lmd.frame_size = frame_size
  end

  # Aジスタを割り当てられるなら割り当てる
  def self.allocate_a( lmd, register_vars )
    a_vars = [] # Aレジスタを割り当てる変数
    register_vars.each do |v,info|
      next unless v.type.size == 1
      if v.live_range.max-v.live_range.min == 1
        # Aレジスタを割り当てられる組み合わせでなければスルー
        op = lmd.ops[v.live_range.min]
        next unless op[1] == v
        next unless [:load, :add, :sub, :and, :or, :xor, 
                     :mul, :div, :mod, :uminus, :eq, :lt, :pget].include?( op[0] )

        next_op = lmd.ops[v.live_range.min+1]
        case next_op[0]
        when :load, :sign_extension, :add, :and, :or, :xor, :eq, :lt, :pget, :sub, :push_arg
          next unless next_op[2] == v
        when :if, :return
          next unless next_op[1] == v
        else next
        end

        a_vars << v
      end
    end
    a_vars.each do |v|
      v.location = :a
      register_vars.delete(v)
    end
  end

  # コンディションレジスタを割り当てられるなら割り当てる
  def self.allocate_cond( lmd, register_vars )
    cond_vars = [] # コンディションレジスタを割り当てる変数
    register_vars.each do |v,info|
      next unless v.type.size == 1
      if v.live_range.max-v.live_range.min == 1
        # Aレジスタを割り当てられる組み合わせでなければスルー
        op = lmd.ops[v.live_range.min]
        next unless op[1] == v
        next unless [:eq, :lt, :not].include?( op[0] )

        next_op = lmd.ops[v.live_range.min+1]
        case next_op[0]
        when :'if'
          next unless next_op[1] == v
        when :'not'
          next unless next_op[2] == v
        else next
        end

        case op[0]
        when :eq 
          v.location = :cond
          v.cond_positive = true
          v.cond_reg = :zero
        when :lt
          v.location = :cond
          v.cond_positive = true
          if op[2].type.signed or op[3].type.signed
            next if op[2].type.size > 1 or op[3].type.size > 1 # サイズ2以上の符号付き比較はフラグが特定できない
            v.cond_reg = :negative
          else
            v.cond_reg = :carry
          end
        when :'not'
          next unless op[2].location == :cond
          v.location = :cond
          v.cond_reg = op[2].cond_reg
          v.cond_positive = !op[2].cond_positive
        else raise
        end
        cond_vars << v
      end
    end
    cond_vars.each do |v|
      #v.location = :a
      register_vars.delete(v)
    end
  end

  ######################################################################
  # レジスタアロケータ
  ######################################################################
  class Allocator

    attr_reader :regs

    def initialize( vars )
      # レジスタの割り当て
      @regs = []
      vars.each do |id,var|
        found = false
        @regs.each do |reg|
          unless overlap?( reg[:live_range], var[:live_range] )
            reg[:live_range] = join( reg[:live_range], var[:live_range] )
            reg[:vars] << id
            found = true
            break
          end
        end
        unless found
          @regs << { live_range: var[:live_range], vars: [id] }
        end
      end
    end

    private

    def overlap?( r1, r2 )
      r1.max >= r2.min and r1.min <= r2.max
    end

    def join( r1, r2 )
      ( [r1.min,r2.min].min .. [r1.max,r2.max].max )
    end


  end

  ######################################################################
  # Live Range の計算を行うクラス
  ######################################################################
  class LiveRangeCalculator

    # flow: 制御フローグラフ( [節0情報,節1情報, ... ] の形で、節情報 = [ [後続節番号,...], [先行節番号,...] ] )
    def initialize( flow )
      @flow = flow.map.with_index do |node,i|
        node << i+1 if i < flow.size-1 # 直後の節を後続節とする
        [node,[]] 
      end

      @flow.each_with_index do |node,i|
        node[0].each do |succ|
          @flow[succ][1] << i
        end
      end

    end

    # live range を計算する
    # 全く使われていない場合、nilを返す
    def calc_live_range( var )
      lives = @flow.map { |i| live = Hash.new }
      var[0].each { |i| lives[i][:define] = true }
      var[1].each { |i| lives[i][:use] = true; lives[i][:live_in] = true }
      while true
        finished = true
        (lives.size-1).downto(0) do |i| 
          live = lives[i]
          # 入り口生存なら、先行節で出口生存
          if live[:live_in]
            @flow[i][1].each do |pred|
              unless lives[pred][:live_out]
                lives[pred][:live_out] = true
                finished = false
              end
            end
          end
          # 出口生存で定義節でないなら、入り口生存
          if live[:live_out] and !live[:define]
            unless live[:live_in]
              live[:live_in] = true
              finished = false
            end
          end
        end
        break if finished
      end
      # show_lives lives
      # live range の算出
      max = 0
      min = 1000000
      lives.each_with_index do |live,i|
        if live[:live_in] or live[:live_out]
          max = i if i > max
          min = i if i < min
        end
      end
      if min == 1000000
        nil
      else
        (min..max)
      end
    end
  end

  #:nocov:
  def show_lives( lives )
    t = { true =>'x', false=>' ', nil=>' ' }
    puts '   n U D  i o'
    lives.each_with_index do |live,i|
        puts( '%04d %s %s  %s %s' % [i, t[live[:use]], t[live[:define]], t[live[:live_in]], t[live[:live_out]]] )
    end
  end
  #:nocov:

end

File: internal/fc/allocator.go
package fc

// レジスタ割付。lib/fc/allocator.rb 由来。
// use_define / register_vars は Ruby の Hash と同じく挿入順を保つ必要がある。
// CastedValue は Delegator のため下位の Value と同一キーに合流する (UnderlyingValue)。

import "fmt"

// ---------------------------------------------------------------
// live range の計算
// ---------------------------------------------------------------

type useDefineEntry struct {
	v       *Value
	defines []int
	uses    []int
}

// CalcLiveRange は Fc.calc_live_range 相当。lmd の各ローカル変数に live_range を設定する。
func CalcLiveRange(lmd *Lambda) {
	// ラベルの収集
	labels := map[string]int{}
	for i, op := range lmd.Ops {
		if op != nil && op.Code == OpLabel {
			labels[op.Label] = i
		}
	}

	// 変数の定義・使用、制御フローグラフの集計
	var udOrder []*useDefineEntry
	udIndex := map[*Value]*useDefineEntry{}
	record := func(vAny Operand, isDefine bool, i int) {
		if vAny == nil {
			return
		}
		if ValKind(vAny) != KindLocal {
			return
		}
		v := UnderlyingValue(vAny)
		if v == nil {
			// PointeredArray がローカル変数を包む場合、Ruby版は live_range 設定で
			// NoMethodError になる (実際には発生しない経路)
			panic(fmt.Sprintf("cannot record %T in use_define", vAny))
		}
		e := udIndex[v]
		if e == nil {
			e = &useDefineEntry{v: v}
			udIndex[v] = e
			udOrder = append(udOrder, e)
		}
		if isDefine {
			e.defines = append(e.defines, i)
		} else {
			e.uses = append(e.uses, i)
		}
	}

	flow := make([][]int, 0, len(lmd.Ops))
	for i, op := range lmd.Ops {
		var uses, defines []Operand
		var node []int
		switch op.Code {
		case OpLabel, OpAsm, OpPushResult, OpPushFastcallResult:
			// DO NOTHING
		case OpIf:
			uses = append(uses, op.Src[0])
			node = append(node, labels[op.Label])
		case OpJump:
			node = append(node, labels[op.Label])
		case OpReturn:
			uses = append(uses, op.src(0))
		case OpPushArg, OpPushFastcallArg:
			uses = append(uses, op.Src[0])
		case OpLoad, OpUminus, OpNot, OpSignExtension, OpRef, OpCall, OpFastcall:
			defines = append(defines, op.Dst)
			uses = append(uses, op.Src[0])
		case OpAdd, OpSub, OpAnd, OpOr, OpXor,
			OpMul, OpDiv, OpMod, OpEq, OpLt,
			OpShiftLeft, OpShiftRight, OpIndex, OpPget:
			defines = append(defines, op.Dst)
			uses = append(uses, op.Src...)
		case OpPset:
			uses = append(uses, op.Src[0], op.Src[1])
		default:
			panic(fmt.Sprintf("invalid op %v", dumpOp(op, nil)))
		}

		flow = append(flow, node)

		for _, v := range defines {
			record(v, true, i)
		}
		for _, v := range uses {
			record(v, false, i)
		}
	}

	// 引数は、最初に定義されているものとする
	for _, e := range udOrder {
		if eqAny(e.v.Opt.GetOr(Sym("local_type")), Sym("arg")) {
			e.defines = append(e.defines, 0)
		}
	}

	// live range を求める
	lrc := NewLiveRangeCalculator(flow)
	for _, e := range udOrder {
		e.v.LiveRange = lrc.CalcLiveRange(e.defines, e.uses)
	}
}

// ---------------------------------------------------------------
// レジスタ割付
// ---------------------------------------------------------------

// AllocateRegister は Fc.allocate_register 相当。
// 変数の address, location, unuse が設定される。
func AllocateRegister(lmd *Lambda) {
	CalcLiveRange(lmd)

	// ref(&演算子)を受けた変数を集める
	refered := map[*Value]bool{}
	for _, op := range lmd.Ops {
		if op != nil && op.Code == OpRef {
			refered[UnderlyingValue(op.Src[0])] = true
		}
	}

	// live range を求め、{callをまたぐ|refを受ける|引数}だったら、フレームスタック上に確保する
	frameSize := lmd.Type.Base.Size // 帰り値分を予約しておく
	var registerVars []*allocEntry  // 挿入順を保つ (Ruby の Hash 相当)
	fastcall := lmd.Type.Fastcall()
	for _, v := range lmd.Vars {
		beyondCall := false
		if v.LiveRange != nil {
			for i := v.LiveRange.Min + 1; i <= v.LiveRange.Max-1; i++ {
				if i >= 0 && i < len(lmd.Ops) && lmd.Ops[i] != nil &&
					lmd.Ops[i].Code == OpCall && !ValType(lmd.Ops[i].Src[0]).Fastcall() {
					beyondCall = true
				}
			}
		}

		if eqAny(v.Opt.GetOr(Sym("local_type")), Sym("result")) {
			// 返り値
			if fastcall {
				lmd.Result.Location = LocFastcallReg
			} else {
				lmd.Result.Location = LocFrame
			}
			lmd.Result.Address = 0
		} else if beyondCall || eqAny(v.Opt.GetOr(Sym("local_type")), Sym("arg")) || refered[v] {
			// 引数か、関数をまたいでいるなら、フレームに割り当てる
			v.Address = frameSize
			if fastcall {
				v.Location = LocFastcallReg
			} else {
				v.Location = LocFrame
			}
			frameSize += v.Type.Size
		} else if v.LiveRange != nil {
			// それ以外の使われてる変数は、レジスターメモリに割り当てる
			registerVars = append(registerVars, &allocEntry{key: v, liveRange: v.LiveRange})
		} else {
			// 未使用フラグをたてる
			v.Location = LocUnused
			v.Unuse = true
		}
	}

	registerVars = allocateCond(lmd, registerVars)
	registerVars = allocateA(lmd, registerVars)

	// 各レジスタのアドレスを割り当てる
	regSize := 0
	allocator := NewAllocator(registerVars)
	for _, reg := range allocator.Regs {
		// アドレスを算出する
		if regSize > 16 {
			panic(&CompileError{Msg: fmt.Sprintf("frame size over on %s", lmd)})
		}
		// 割り当てる
		for _, kAny := range reg.vars {
			v := kAny.(*Value)
			if fastcall {
				if frameSize+regSize > 16 {
					panic(&CompileError{Msg: fmt.Sprintf("frame size over on %s", lmd)})
				}
				v.Location = LocFastcallReg
				v.Address = frameSize + regSize
			} else {
				v.Location = LocReg
				v.Address = regSize
			}
		}
		regSize += 2
	}

	lmd.FrameSize = frameSize
}

// allocateA は Aレジスタを割り当てられるなら割り当てる。
func allocateA(lmd *Lambda, registerVars []*allocEntry) []*allocEntry {
	var aVars []*Value
	for _, e := range registerVars {
		v := e.key
		if v.Type.Size != 1 {
			continue
		}
		if v.LiveRange.Max-v.LiveRange.Min == 1 {
			// Aレジスタを割り当てられる組み合わせでなければスルー
			op := lmd.Ops[v.LiveRange.Min]
			if !isSameValue(op.Dst, v) {
				continue
			}
			if !codeIn(op.Code, OpLoad, OpAdd, OpSub, OpAnd, OpOr, OpXor,
				OpMul, OpDiv, OpMod, OpUminus, OpEq, OpLt, OpPget) {
				continue
			}

			nextOp := lmd.Ops[v.LiveRange.Min+1]
			switch nextOp.Code {
			case OpLoad, OpSignExtension, OpAdd, OpAnd, OpOr, OpXor,
				OpEq, OpLt, OpPget, OpSub, OpPushArg,
				OpIf, OpReturn:
				// 最初の入力オペランドが v であること (旧実装の op[2] / if・return では op[1] に相当)
				if !isSameValue(nextOp.src(0), v) {
					continue
				}
			default:
				continue
			}

			aVars = append(aVars, v)
		}
	}
	for _, v := range aVars {
		v.Location = LocA
		registerVars = deleteEntry(registerVars, v)
	}
	return registerVars
}

// allocateCond はコンディションレジスタを割り当てられるなら割り当てる。
func allocateCond(lmd *Lambda, registerVars []*allocEntry) []*allocEntry {
	var condVars []*Value
	for _, e := range registerVars {
		v := e.key
		if v.Type.Size != 1 {
			continue
		}
		if v.LiveRange.Max-v.LiveRange.Min == 1 {
			op := lmd.Ops[v.LiveRange.Min]
			if !isSameValue(op.Dst, v) {
				continue
			}
			if !codeIn(op.Code, OpEq, OpLt, OpNot) {
				continue
			}

			nextOp := lmd.Ops[v.LiveRange.Min+1]
			switch nextOp.Code {
			case OpIf, OpNot:
				if !isSameValue(nextOp.src(0), v) {
					continue
				}
			default:
				continue
			}

			switch op.Code {
			case OpEq:
				v.Location = LocCond
				v.CondPositive = true
				v.CondReg = CondZero
			case OpLt:
				v.Location = LocCond
				v.CondPositive = true
				if ValType(op.Src[0]).Signed || ValType(op.Src[1]).Signed {
					if ValType(op.Src[0]).Size > 1 || ValType(op.Src[1]).Size > 1 {
						// サイズ2以上の符号付き比較はフラグが特定できない。
						// Ruby版は next の前に location/cond_positive を設定済みのまま残す
						// (後で register_vars の割付により location は上書きされる)
						continue
					}
					v.CondReg = CondNegative
				} else {
					v.CondReg = CondCarry
				}
			case OpNot:
				if ValLocation(op.Src[0]) != LocCond {
					continue
				}
				v.Location = LocCond
				v.CondReg = UnderlyingValue(op.Src[0]).CondReg
				v.CondPositive = !UnderlyingValue(op.Src[0]).CondPositive
			default:
				panic("unreachable")
			}
			condVars = append(condVars, v)
		}
	}
	for _, v := range condVars {
		registerVars = deleteEntry(registerVars, v)
	}
	return registerVars
}

func isSameValue(opElem Operand, v *Value) bool {
	// Ruby の op[1] == v (CastedValue は Delegator の == で from と比較される)
	return opElem != nil && UnderlyingValue(opElem) == v
}

func codeIn(c OpCode, codes ...OpCode) bool {
	for _, x := range codes {
		if c == x {
			return true
		}
	}
	return false
}

func deleteEntry(entries []*allocEntry, v *Value) []*allocEntry {
	r := entries[:0]
	for _, e := range entries {
		if e.key != v {
			r = append(r, e)
		}
	}
	return r
}

// ---------------------------------------------------------------
// レジスタアロケータ
// ---------------------------------------------------------------

type allocEntry struct {
	key       *Value
	liveRange *LiveRange
}

type AllocatorReg struct {
	liveRange *LiveRange
	vars      []any
}

type Allocator struct {
	Regs []*AllocatorReg
}

func NewAllocator(vars []*allocEntry) *Allocator {
	a := &Allocator{}
	for _, e := range vars {
		found := false
		for _, reg := range a.Regs {
			if !overlapRange(reg.liveRange, e.liveRange) {
				reg.liveRange = joinRange(reg.liveRange, e.liveRange)
				reg.vars = append(reg.vars, e.key)
				found = true
				break
			}
		}
		if !found {
			a.Regs = append(a.Regs, &AllocatorReg{liveRange: e.liveRange, vars: []any{e.key}})
		}
	}
	return a
}

// NewAllocatorGeneric はキーが任意の値のバージョン (単体テスト用)。
func NewAllocatorGeneric(keys []any, ranges []*LiveRange) *Allocator {
	a := &Allocator{}
	for i, k := range keys {
		lr := ranges[i]
		found := false
		for _, reg := range a.Regs {
			if !overlapRange(reg.liveRange, lr) {
				reg.liveRange = joinRange(reg.liveRange, lr)
				reg.vars = append(reg.vars, k)
				found = true
				break
			}
		}
		if !found {
			a.Regs = append(a.Regs, &AllocatorReg{liveRange: lr, vars: []any{k}})
		}
	}
	return a
}

func overlapRange(r1, r2 *LiveRange) bool {
	return r1.Max >= r2.Min && r1.Min <= r2.Max
}

func joinRange(r1, r2 *LiveRange) *LiveRange {
	return &LiveRange{Min: min(r1.Min, r2.Min), Max: max(r1.Max, r2.Max)}
}

// ---------------------------------------------------------------
// Live Range の計算を行うクラス
// ---------------------------------------------------------------

type LiveRangeCalculator struct {
	flow [][2][]int // [後続節, 先行節]
}

// NewLiveRangeCalculator は flow (各opの後続節リスト) から構築する。
// 直後の節は自動的に後続節に追加される (最後の節を除く)。
func NewLiveRangeCalculator(flow [][]int) *LiveRangeCalculator {
	l := &LiveRangeCalculator{flow: make([][2][]int, len(flow))}
	for i, node := range flow {
		succ := append([]int{}, node...)
		if i < len(flow)-1 {
			succ = append(succ, i+1)
		}
		l.flow[i][0] = succ
	}
	for i := range l.flow {
		for _, succ := range l.flow[i][0] {
			l.flow[succ][1] = append(l.flow[succ][1], i)
		}
	}
	return l
}

// CalcLiveRange は live range を計算する。全く使われていない場合 nil を返す。
func (l *LiveRangeCalculator) CalcLiveRange(defines, uses []int) *LiveRange {
	type liveInfo struct {
		define, use, liveIn, liveOut bool
	}
	lives := make([]liveInfo, len(l.flow))
	for _, i := range defines {
		lives[i].define = true
	}
	for _, i := range uses {
		lives[i].use = true
		lives[i].liveIn = true
	}
	for {
		finished := true
		for i := len(lives) - 1; i >= 0; i-- {
			// 入り口生存なら、先行節で出口生存
			if lives[i].liveIn {
				for _, pred := range l.flow[i][1] {
					if !lives[pred].liveOut {
						lives[pred].liveOut = true
						finished = false
					}
				}
			}
			// 出口生存で定義節でないなら、入り口生存
			if lives[i].liveOut && !lives[i].define {
				if !lives[i].liveIn {
					lives[i].liveIn = true
					finished = false
				}
			}
		}
		if finished {
			break
		}
	}
	// live range の算出
	maxI := 0
	minI := 1000000
	for i := range lives {
		if lives[i].liveIn || lives[i].liveOut {
			if i > maxI {
				maxI = i
			}
			if i < minI {
				minI = i
			}
		}
	}
	if minI == 1000000 {
		return nil
	}
	return &LiveRange{Min: minI, Max: maxI}
}

// ---------------------------------------------------------------
// 使っていない変数の削除 (Ruby版では Llc#delete_unuse)
// ---------------------------------------------------------------

func DeleteUnuse(lmd *Lambda) {
	for i, op := range lmd.Ops {
		if op == nil {
			continue
		}
		switch op.Code {
		case OpPget, OpLoad:
			if op.Dst != nil && UnderlyingValue(op.Dst) != nil && UnderlyingValue(op.Dst).Unuse {
				lmd.Ops[i] = nil
			}
		case OpCall, OpFastcall:
			if op.Dst != nil && UnderlyingValue(op.Dst) != nil && UnderlyingValue(op.Dst).Unuse {
				op.Dst = nil
			}
		}
	}
}
